torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -32,6 +32,228 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+    if (has_vectorize_op) {
+      fl->setVectorize(true);
+      loop_structure.push_back(fl);
+      return first_expr;
+    }
+  }
+  return nullptr;
+}
+
+// Get full extent for the inner-most, merged root domain
+kir::Val* getVectorizeExtent(kir::TensorIndex* node) {
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  auto domain = node->view()->domain()->rootDomain();
+  auto root_domain_index_map = node->rootDomainIndexMap();
+
+  auto extent = domain.back()->extent();
+  if (domain.size() >= 2) {
+    for (int64_t i = domain.size() - 2; i >= 0; --i) {
+      if (!root_domain_index_map[i]) {
+        extent = ir_builder.mulExpr(extent, domain[i]->extent());
+      } else {
+        break;
+      }
+    }
+  }
+  return extent;
+}
+
+// Generate Index Value
+kir::Val* generateBaseIndex(kir::TensorIndex* node) {
+  if (node->indices().size() == 1) {
+    return node->indices().front();
+  }
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  auto result = node->indices().front();
+  for (size_t idx = 1; idx < node->indices().size(); ++idx) {
+    result = ir_builder.addExpr(result, node->indices()[idx]);
+  }
+  return result;
+}
+
+kir::Val* setupNamedScalar(
+    kir::Scope& body,
+    kir::Val* val,
+    const std::string& name) {
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  kir::Int* one = ir_builder.create<kir::Int>(1);
+
+  auto namedScalar = ir_builder.namedSetExpr(name, val);
+  auto alloc =
+      ir_builder.create<kir::Allocate>(namedScalar, MemoryType::Local, one);
+  body.push_back(alloc);
+  body.push_back(namedScalar->definition());
+
+  return namedScalar;
+}
+
+kir::ForLoop* handleMisalignedVectorization(
+    std::vector<kir::ForLoop*> loop_structure,
+    const kir::ForLoop* for_loop) {
+  // body -> allocate, read, compute, write for-loops
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+  kir::Int* zero = ir_builder.create<kir::Int>(0);
+
+  // create new base For-Loop
+  const auto new_loop = ir_builder.create<kir::ForLoop>(
+      for_loop->index(), for_loop->extent(), for_loop->iter_domain());
+
+  // Find child For-Loops and add remaining expressions to base For-Loop
+  auto child_loops = parseVectorizedForLoop(for_loop, new_loop);
+
+  // Find the first vectorize set - either read or write
+  // Add child For-Loop to loop_structure
+  // Enable vectorize flag in child For-Loop
+  auto vec_expr = findVectorizedSet(loop_structure, child_loops);
+  TORCH_INTERNAL_ASSERT(vec_expr != nullptr);
+  TORCH_INTERNAL_ASSERT(vec_expr->outputs().front()->isA<kir::TensorView>());
+  TORCH_INTERNAL_ASSERT(vec_expr->inputs().front()->isA<kir::TensorView>());
+
+  // out_tv is the TensorView with the misaligned vec iterDomain
+  auto out_tv = vec_expr->outputs().front()->as<kir::TensorView>();
+  auto in_tv = vec_expr->inputs().front()->as<kir::TensorView>();
+  kir::TensorView* vec_tv =
+      (out_tv->memoryType() == MemoryType::Local) ? out_tv : in_tv;
```

Comment:
I'm guess here that you are trying to find whether this should be vectorized store or load, and if it's store, the producer tensor must be on local memory, and if it's load, the consumer must be local memory.

What if we want to do vectorized-load to shared memory? What about vectorized store from shared memory?

Reply:
Here I'm looking for the TensorView with the Misaligned Vectorization axis.

`out_tv->memoryType() != MemoryType::Global` is a better choice.

Reply:
I've seen vectorized load from shared to registers. 

Let's focus on global <-> register vectorization for now. Just please make sure erroring out for other cases.

Reply:
How do you propose to handle the shift in shared memory? The shared memory access need to be aligned as well.

Reply:
Vectorized load from shared to local should look similar to vectorized load from global to local. 

Vectorized load from global to shared seems tricky. We should probably just use the memcpy_async intrinsic.

Reply:
Vectorized loads from smem are most urgently needed aligned. Unclear if there are any use cases of misaligned for smem.

The only reason we do vectorized loads from smem in practice is for gemm/conv which uses aligned vectorization.

Might be some uses some day, but definitely not an urgent concern.